### PR TITLE
Criteo Bid Adapter: add iframe syncs support

### PR DESCRIPTION
**Pull request by afewcc**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

Added support for user sync through iframe as possible sync option

## Contacts
- Tachfine El Belky (t.elbelky@criteo.com)

### CODE REVIEW

**Comment by ChrisHuie** [comment]
> Hello @ChrisHuie I hope your are doing well :) Any chance to get this one reviewed in the coming days ? Thanks a lot

Should be able to review it in the next couple of days

File: modules/criteoBidAdapter.js [hunk 1]
```diff
@@ -31,12 +32,108 @@ const PUBLISHER_TAG_URL_TEMPLATE = 'https://static.criteo.net/js/ld/publishertag
...
+        origin: origin,
+        requestId: requestId,
+        secureIdCookie: readFromAllStorages(SID_COOKIE_NAME),
+        tld: refererInfo.domain,
+        topUrl: refererInfo.domain,
+        uid: readFromAllStorages(IDCPY_COOKIE_NAME),
+        version: '$prebid.version$'.replace(/\./g, '_'),
+      };
+
+      window.addEventListener('message', function (event) {
```
Comment (dgirardi) [inline modules/criteoBidAdapter.js:91]: this leaves a dangling listener each time it's run; imo it should detach itself when it gets the matching requestId back

File: modules/criteoBidAdapter.js [hunk 2]
```diff
@@ -209,6 +306,34 @@ export const spec = {
...
+  storage.setCookie(name, value, expires);
+  storage.setDataInLocalStorage(name, value);
+}
+
+function deleteFromAllStorages(name) {
+  storage.setCookie(name, '', 0);
+  storage.removeDataFromLocalStorage(name);
+}
+
+function hasPurpose1Consent(gdprConsent) {
```
Comment (dgirardi) [inline modules/criteoBidAdapter.js:330]: You could use this utility: https://github.com/prebid/Prebid.js/blob/0be4560e182e46c121c283f763cdef74b5ff9db9/src/utils/gpdr.js#L9

File: modules/criteoBidAdapter.js [hunk 3]
```diff
@@ -31,12 +32,108 @@ const PUBLISHER_TAG_URL_TEMPLATE = 'https://static.criteo.net/js/ld/publishertag
...
+  getUserSyncs: function (syncOptions, _, gdprConsent, uspConsent) {
+    const fastBidVersion = config.getConfig('criteo.fastBidVersion');
+    if (canFastBid(fastBidVersion)) {
+      return [];
+    }
+
+    const refererInfo = getRefererInfo();
+    const origin = 'criteoPrebidAdapter';
+
+    if (syncOptions.iframeEnabled && hasPurpose1Consent(gdprConsent)) {
```
Comment (patmmccann) [inline modules/criteoBidAdapter.js:58]: where is iframeEnabled coming from 

https://github.com/prebid/Prebid.js/blob/0be4560e182e46c121c283f763cdef74b5ff9db9/modules/pubmaticBidAdapter.js#L1300 is a more typical use of iframe syncs

Comment (patmmccann) [reply modules/criteoBidAdapter.js:58]: i see, you are trying to decorate the sync call with these id's.

Can you please indicate in your docs that 'opting into iframe syncs with this bidder may set values in the first party context'

File: modules/criteoBidAdapter.js (hunk 1 above)
Comment (dgirardi) [reply modules/criteoBidAdapter.js:91]: I'm also curious: how does the sync figure out which window to message? this is the first time I see this approach and I'm wondering if it needs support from core for the general case (where prebid might be in a window other than `top`).

Comment (leonardlabat) [reply modules/criteoBidAdapter.js:91]: From the sync frame, we simply try to send the message to window.parent (which should be the one listening to the message as it's the one which dropped the frame)

File: modules/criteoBidAdapter.js (hunk 2 above)
Comment (leonardlabat) [reply modules/criteoBidAdapter.js:330]: done

File: modules/criteoBidAdapter.js (hunk 3 above)
Comment (leonardlabat) [reply modules/criteoBidAdapter.js:58]: Regarding the _hasPurpose1Consent_ call
We really want the purpose1 to be granted prior to init our user sync as it involves interacting with 1PC the LS to resolve our sync endpoint arguments.

**Review by patmmccann — APPROVED** [review]

**Pull request merged**